### PR TITLE
Go based validator

### PR DESCRIPTION
**Pull request by TAvardit**
This PR includes:
 - Go app + unit tests
 - Docker file to build and run app publicly

### CODE REVIEW

File: util/main.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	}, []string{"code", "method"})
+
+	httpRequestDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "http_request_duration_seconds",
+		Help: "Duration of all HTTP requests",
+	}, []string{"code", "handler", "method"})
+)
+
+func apiGet(url string) ([]byte, error) {
+	client := &http.Client{}
```
[the remark this answers is not in the document]
Comment (TAvardit) [reply util/main.go:85]: Great comment! I sweated on making it work but now it is!

File: util/main.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return body, nil
+
+}
+
+func clear(v interface{}) {
```
[the remark this answers is not in the document]
Comment (TAvardit) [reply util/main.go:108]: Will improve naming :)

File: Dockerfile [hunk 3]
```diff
@@ -3,19 +3,22 @@ WORKDIR  /go/src/github.com/travelaudience/prometheus-validator
...
 EXPOSE 8080
 
 
-COPY --from=build /go/src/github.com/travelaudience/prometheus-validator/util /
+COPY --from=build /go/src/github.com/travelaudience/prometheus-validator/util/util /
 
-# ENV PROMETHEUS_RULES_API_URL="http://localhost:9090/api/v1/rules?type=alert"
-ENV PROMETHEUS_RULES_API_URL="http://prometheus-operated.monitoring.svc.cluster.local:9090/api/v1/rules?type=alert"
+# ENV PROMETHEUS_URL="http://localhost:9090"
+ENV PROMETHEUS_URL="http://prometheus-operated.monitoring.svc.cluster.local:9090"
```
Comment (psychopenguin) [inline Dockerfile:21]: as the idea for it is to run as sidecar, i think keep it pointing to localhost is a better default value.

File: util/main.go [hunk 4]
```diff
@@ -13,6 +13,10 @@ import (
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
+type MyHTTPClient struct {
```
Comment (psychopenguin) [inline util/main.go:16]: name it something as `prometheusClient`

File: util/main.go [hunk 5]
```diff
@@ -156,7 +159,10 @@ func recordMetrics(queryInterval time.Duration, url string, noPlaybookAlerts *[]
 
 func main() {
 	var noPlaybookAlerts []AlertRule
-	go recordMetrics(1, os.Getenv("PROMETHEUS_RULES_API_URL"), &noPlaybookAlerts)
+	client := &MyHTTPClient{
+		&http.Client{},
+	}
+	go recordMetrics(1, os.Getenv("PROMETHEUS_URL")+"/api/v1/rules?type=alert", &noPlaybookAlerts, client)
```
Comment (psychopenguin) [inline util/main.go:165]: this should work, but i think it's safer to rely on the [url module](https://golang.org/pkg/net/url/) to build the url to be queried, and also, this query should be inside the client.

File: util/main.go [hunk 6]
```diff
@@ -129,8 +132,8 @@ func checkAlerts(apiResp []byte, noPlaybookAlerts *[]AlertRule) error {
 	return nil
 }
 
-func recordMetrics(queryInterval time.Duration, url string, noPlaybookAlerts *[]AlertRule) {
-	apiResp, err := apiGet(url)
+func recordMetrics(queryInterval time.Duration, url string, noPlaybookAlerts *[]AlertRule, client *MyHTTPClient) {
```
Comment (psychopenguin) [inline util/main.go:135]: shouldn't the prometheus client be the receiver for this?

File: Dockerfile (hunk 3 above)
Comment (TAvardit) [reply Dockerfile:21]: NP

File: util/main.go (hunk 6 above)
Comment (TAvardit) [reply util/main.go:135]: Not in my opinion. PrometheusClient is passed in as a pointer so `apiGet` function will use it as a receiver to make api calls to Prometheus.
`recordMetrics` is serving the metrics with Prometheus HTTP handler. or am I getting this all wrong?

File: util/main.go (hunk 5 above)
Comment (TAvardit) [reply util/main.go:165]: Not sure what you mean. The url is being queried in another function.

File: util/main.go (hunk 6 above)
Comment (psychopenguin) [reply util/main.go:135]: IMHO, `apiGet` also should be a method of `prometheusClient` - ie, be the receiver of the fuction :)

**Pull request merged**